Sources/GRPCCore/Transport/InProcessClientTransport.swift

```diff
@@ -0,0 +1,350 @@
+  /// maintains connections. The function exits when all open streams have been closed and new connections
+  /// are no longer required by the caller who signals this by calling ``close()``, or by cancelling the
+  /// task this function runs in.
+  ///
+  /// - Parameter lazily: This parameter is ignored in this implementation.
+  public func connect(lazily: Bool) async throws {
+    let (stream, continuation) = AsyncStream<Void>.makeStream()
+    try self.state.withLockedValue { state in
+      switch state {
+      case .unconnected(let unconnectedState):
+        state = .connected(
+          .init(
+            fromUnconnected: unconnectedState,
+            signalEndContinuation: continuation
+          )
+        )
+        for pendingStream in unconnectedState.pendingStreams {
+          pendingStream.finish()
+        }
+      case .connected:
+        throw RPCError(
+          code: .failedPrecondition,
+          message: "Already connected to server."
+        )
+      case .closed:
+        throw RPCError(
+          code: .failedPrecondition,
+          message: "Can't connect to server, transport is closed."
+        )
+      }
+    }
+
+    for await _ in stream {
+      // This for-await loop will exit (and thus `connect(lazily:)` will return)
+      // only when the task is cancelled, or when the stream's continuation is
+      // finished - whichever happens first.
+      // The continuation will be finished when `close()` is called and there
+      // are no more open streams.
+    }
+
+    // If at this point there are any open streams, it's because Cancellation
+    // occurred and all open streams must be closed at this point.
+    self.state.withLockedValue { state in
+      switch state {
+      case .unconnected:
+        ()
+      case .connected(let connectedState):
+        state = .closed(.init())
+        for (clientStream, serverStream) in connectedState.openStreams.values {
+          clientStream.outbound.finish()
+          serverStream.outbound.finish()
+        }
+      case .closed(let closedState):
+        for (clientStream, serverStream) in closedState.openStreams.values {
+          clientStream.outbound.finish()
+          serverStream.outbound.finish()
+        }
+      }
+    }
+  }
+
+  /// Signal to the transport that no new streams may be created.
+  ///
+  /// Existing streams may run to completion naturally but calling ``withStream(descriptor:_:)``
+  /// will result in an ``RPCError`` with code ``RPCError/Code/failedPrecondition`` being thrown.
+  ///
+  /// If you want to forcefully cancel all active streams then cancel the task running ``connect(lazily:)``.
+  public func close() {
+    self.state.withLockedValue { state in
+      switch state {
+      case .unconnected:
+        state = .closed(.init())
+      case .connected(let connectedState):
+        if connectedState.openStreams.count == 0 {
+          connectedState.signalEndContinuation.finish()
```

Comment:
Is this because it can be a pretty slow operation and we try to have as small a critical region as possible, or something else?

Reply:
Partially to minimise the critical region but mostly because making out calls (i.e. anything which could cause unknown side effects) can be very difficult to reason about. The recent calling continuations while holding a lock is an example of this: we _thought_ it was a safe outcall to make but ended up causing weird deadlock bugs. It's best to just avoid it entirely.